### PR TITLE
ENH Kymatio Frontends [2]

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
As discussed in #432 , it's better to split the pull request into smaller pull request. This pull request adds the new frontend classes, which will be used in 0.2.0. 

Also see #440 .

Recommended to merge 2nd.

### CODE REVIEW

**Comment by edouardoyallon** [comment]
@MuawizChaudhary maybe you can incorporate your frontend here? Also, @janden wants something without the "ABC", yet instead, raise NotImplementedError. The only thing it will change is that:

```
class ScatteringTorch(nn.Module, metaclass=ABCMeta):
```

becomes

```
class ScatteringTorch(nn.Module, ScatteringGeneral):
```

We remove the `import` relative to `abc` and for instance:

```
   @abstractmethod
   def scattering(self, x):
       """ This function should compute the scattering transform."""
```

will now be

```
   def scattering(self, x):
       """ This function should compute the scattering transform."""
      raise NotImplementedError
```

That's all.

**Comment by edouardoyallon** [comment]
Also there is not unittest here, that's *fine*.

**Comment by edouardoyallon** [comment]
Hi,

What's the purpose of "super frontend"? beyond that, I'm ok with merging. @eickenberg @janden ??

**Comment by MuawizChaudhary** [comment]
> Hi,
> 
> What's the purpose of "super frontend"? beyond that, I'm ok with merging. @eickenberg @janden ??

Did it based on Dr. Lostanlen's [suggestion](https://github.com/kymatio/kymatio/pull/447#issuecomment-546919256): 
> `General` is not a good name for the abstraction.
> Consider replacing by `Super` or `Interface` or `Parent` or `Base`
> 
> https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)
Think interface would be a better name?

File: requirements.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
-numpy
-scipy
+numpy>=1.7
```
Comment (janden) [inline requirements.txt]: As mentioned in #440, let's drop these requirements for now.

File: kymatio/frontend/torch_frontend.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        """ This function should call the functional scattering."""
+        raise NotImplementedError
+
+    def forward(self, x):
+        return self.scattering(x)
+    
+    def loginfo(self):
+        """ Returns the logging message when the frontend is deployed."""
+        raise NotImplementedError
+
```
Comment (janden) [inline kymatio/frontend/torch_frontend.py]: Remove.

**Comment by MuawizChaudhary** [comment]
> It's not clear to me what the intended structure is here. What's going to inherit from `ScatteringSuper`? Why are certain methods (like `build()`) defined both there and in `ScatteringNumPy`, for example?

Yeah, so my plan was to have a specific ScatteringSuper implementation for each *d front end, as currently the Torch, NumPy, and TensorFlow ends share alot of code (I believe you said this in one of your comments too). Then there would be a ScatteringTorch*d, ScatteringNumPy*d, and ScatteringTensorFlow*d frontend which would subclass the ScatteringSuper*d implementation and the appropriate class in the kymatio base frontend folder. 

Is this fine or would you prefer some other system? Is Super fine or do we want to name it parent or interface?

The build() should not be there. 

You have alot of "remove", but I believe Dr. Oyallon said you wanted to remove the abc structure and have each method raise not implemented errors instead. Correct me if I am mistaken.

File: kymatio/__init__.py [hunk 3]
```diff
@@ -12,8 +13,9 @@
             'HarmonicScattering3D'
             ]
 
-from .scattering1d.scattering1d import Scattering1D
-from .scattering2d.scattering2d import Scattering2D
-from .scattering3d.scattering3d import HarmonicScattering3D
+from .scattering1d import Scattering1D
+from .scattering2d import Scattering2D
+from .scattering3d import HarmonicScattering3D
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/__init__.py]: @edouardoyallon

[the remark this answers is not in the document]
Comment (edouardoyallon) [reply kymatio/__init__.py]: revert: it allows avoid declaring an init file

**Comment by edouardoyallon** [comment]
> On the subject of removing: @edouardoyallon you were the one that added in the loginfo, do you have any arguments for keeping it in? 

 Actually, with inheritance, the idea is to factorize the code. Thus, the "super" frontend should be inherited by torch, np and tf frontend. Then, redundant functions should be avoided: `loginfo` should be defined in the "super" frontend.

I'm ok to remove the loginfo function from this PR and to add it later. It's pretty useful for debugging and tracking purposes.

File: kymatio/frontend/torch_frontend.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+from packaging import version
+
+assert version.parse(torch.__version__)>=version.parse("1.3"), 'Current PyTorch version is ' + str(torch.__version__) +\
+                                                    ' . Please upgrade PyTorch to 1.3 at least.'
+
+class ScatteringTorch(nn.Module):
+    def __init__(self):
+        super(ScatteringTorch, self).__init__()
+    
+    def build(self):
```
Comment (MuawizChaudhary) [inline kymatio/frontend/torch_frontend.py]: Remove build, define register_backend.

File: kymatio/frontend/torch_frontend.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        """ Defines elementary routines. 
+        This function should always call and create the filters via
+        self.create_and_register_filters() defined below. For instance, via:
+        self.filters = self.create_and_register_filters() """
+        raise NotImplementedError
+    
+    def create_filters(self):
```
Comment (MuawizChaudhary) [inline kymatio/frontend/torch_frontend.py]: Define register_filter.

File: kymatio/frontend/base_frontend.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+
+class ScatteringBase():
```
Comment (MuawizChaudhary) [inline kymatio/frontend/base_frontend.py]: Check all current Scattering*D files and make sure there aren't other functions we need to define.

**Comment by MuawizChaudhary** [comment]
> > On the subject of removing: @edouardoyallon you were the one that added in the loginfo, do you have any arguments for keeping it in?
> 
> Actually, with inheritance, the idea is to factorize the code. Thus, the "super" frontend should be inherited by torch, np and tf frontend. Then, redundant functions should be avoided: `loginfo` should be defined in the "super" frontend.
> 

Perhaps I should slow down a bit on making some of these changes. 

In my current implementation (#433) the ScatteringND\*, where * in {torch, numpy, tensorflow}, inherits from both Scattering* and an implementation of ScatteringSuper. Is there some other better way of doing it?

File: kymatio/frontend/base_frontend.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    def __init__(self):
+        super(ScatteringBase, self).__init__()
+    
+    def build(self):
+        """ Defines elementary routines.
```
Comment (MuawizChaudhary) [inline kymatio/frontend/base_frontend.py]: For the doc strings in the front ends section do we want to follow the [Numpy](https://numpydoc.readthedocs.io/en/latest/format.html) doc strings convention?

**Comment by MuawizChaudhary** [comment]
Thinking about it, every frontend registers the backend, across all the *d. Perhaps we should have those implemented in the corresponding torch, numpy, tensorflow frontend defined here?

I'll commit some stuff related to that soon, if they aren't to y'all's liking, we can simply revert those commits.

**Comment by edouardoyallon** [comment]
> Thinking about it, every frontend registers the backend, across all the *d. Perhaps we should have those implemented in the corresponding torch, numpy, tensorflow frontend defined here?
> 
> I'll commit some stuff related to that soon, if they aren't to y'all's liking, we can simply revert those commits.

Tbh I'm gonna pretty easy-going on this one! A later refactoring would be possible and will be plausible: it shouldn't affect much the future user of this new pkg!

File: requirements.txt (hunk 1 above)
Comment (edouardoyallon) [reply requirements.txt]: Feel free @MuawizChaudhary to open an issue!

File: kymatio/frontend/base_frontend.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        save those arrays as module's buffers. """
+        raise NotImplementedError
+
+    def scattering(self, x):
+        """ This function should call the functional scattering."""
+        raise NotImplementedError
+    
+    def loginfo(self):
+        """ Returns the logging message when the frontend is deployed."""
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (edouardoyallon) [reply kymatio/frontend/base_frontend.py]: If Joakim said to rmeove, I'm ok with it. I'll add later the log messages.

**Comment by janden** [comment]
> To elucidate with an example on the 1d branch: I would first implement ScatteringSuper1D, which inherits from ScatteringSuper. Then I would implement ScatteringTorch1D by inheriting from ScatteringSuper1D and from ScatteringTorch, and implementing the required methods.

Ok that could work. What would each of these classes implement (that isn't implemented already in one of the parent classes)? I'm just trying to get a sense of what the division of responsibility is here.

> I'll change the name from ScatteringSuper to ScatteringBase, unless you believe strongly in just Scattering.

It might be good to reserve the bare `Scattering` to the user-facing classes (like `Scattering1D`), so `ScatteringBase` might be better here.

> Thinking about it, every frontend registers the backend, across all the *d. Perhaps we should have those implemented in the corresponding torch, numpy, tensorflow frontend defined here?

I don't understand. What does “registering” mean here?

File: kymatio/frontend/base_frontend.py (hunk 7 above)
Comment (janden) [reply kymatio/frontend/base_frontend.py]: We should follow whatever convention is used by the other user-facing functions. I believe it's the same as NumPy, but I'm not 100% certain.

File: kymatio/frontend/base_frontend.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (janden) [reply kymatio/frontend/base_frontend.py]: Yes I'm not sure why we need to reimplement the logging package [that comes standard with Python](https://docs.python.org/3/library/logging.html).

**Comment by MuawizChaudhary** [comment]
> > To elucidate with an example on the 1d branch: I would first implement ScatteringSuper1D, which inherits from ScatteringSuper. Then I would implement ScatteringTorch1D by inheriting from ScatteringSuper1D and from ScatteringTorch, and implementing the required methods.
> 
> Ok that could work. What would each of these classes implement (that isn't implemented already in one of the parent classes)? I'm just trying to get a sense of what the division of responsibility is here.
> 
ScatteringSuper1D would implement the functions in ScatteringSuper. ScatteringTorch1D would implement the functions in ScatteringTorch. Let me know if you need more detail.

> > I'll change the name from ScatteringSuper to ScatteringBase, unless you believe strongly in just Scattering.
> 
> It might be good to reserve the bare `Scattering` to the user-facing classes (like `Scattering1D`), so `ScatteringBase` might be better here.
> 

I believe I've made this change.

**Comment by janden** [comment]
> I believe I've made this change.

Right. I didn't see. Sorry about that.

File: kymatio/frontend/base_frontend.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    def build(self):
+        """ Defines elementary routines. 
+
+        This function should always call and create the filters via
+        self.create_filters() defined below. For instance, via:
+        self.filters = self.create_filters() """
+        raise NotImplementedError
+    
+    def register_backend(self, string):
+        """ This function should register the backend to be used"""
```
[the remark this answers is not in the document]
Comment (MuawizChaudhary) [reply kymatio/frontend/base_frontend.py]: As discussed on the slack, we'll change name from `register_backend` to `set_backend`

File: kymatio/frontend/base_frontend.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        """ This function should instantiate the backend to be used if not already
+        specified"""
+        if isinstance(self.backend, str):
+            self.backend = importlib.import_module(import_string + self.backend \
+                    + "_backend", 'backend')
+        elif not self.backend.name.startswith(self.frontend_name):
+            raise RuntimeError('This backend is not supported.')
+
+ 
+
```
Comment (MuawizChaudhary) [inline kymatio/frontend/base_frontend.py]: Is it me or is this a lot of space?

Comment (janden) [reply kymatio/frontend/base_frontend.py]: It is.

**Comment by janden** [comment]
Ok. We're good here, as far as I'm concerned.

@edouardoyallon, @eickenberg?

**Comment by edouardoyallon** [comment]
> Ok. We're good here, as far as I'm concerned.
> 
> @edouardoyallon, @eickenberg?

Please merge.

**Pull request merged**